packages/back-end/src/models/ExperimentModel.ts

```diff
@@ -107,18 +121,377 @@ const experimentSchema = new mongoose.Schema({
+        found = key;
+      }
+      n++;
+    }
+
+    // Fall back to uniqid if couldn't generate
+    data.trackingKey = found || uniqid();
+  }
+
+  const nextUpdate = determineNextDate(
+    organization.settings?.updateSchedule || null
+  );
+
+  const exp = await ExperimentModel.create({
+    ...data,
+    id: uniqid("exp_"),
+    dateCreated: new Date(),
+    dateUpdated: new Date(),
+    autoSnapshots: nextUpdate !== null,
+    lastSnapshotAttempt: new Date(),
+    nextSnapshotAttempt: nextUpdate,
+  });
+
+  if (data.tags) {
+    await addTags(data.organization, data.tags);
+  }
+
+  return toInterface(exp);
+}
+
+export async function updateExperimentById(
+  organization: string,
+  experiment: ExperimentInterface,
+  changes: Changeset
+): Promise<void> {
+  await ExperimentModel.updateOne(
+    {
+      id: experiment.id,
+      organization,
+    },
+    {
+      $set: changes,
+    }
+  );
+}
+
+export async function getExperimentsByMetric(
+  organization: string,
+  metricId: string
+): Promise<{ id: string; name: string }[]> {
+  const experiments: { id: string; name: string }[] = [];
+
+  const cols = {
+    _id: false,
+    id: true,
+    name: true,
+  };
+
+  // Using as a goal metric
+  const goals = await ExperimentModel.find(
+    {
+      organization,
+      metrics: metricId,
+    },
+    cols
+  );
+  goals.forEach((exp) => {
+    experiments.push({
+      id: exp.id,
+      name: exp.name,
+    });
+  });
+
+  // Using as a guardrail metric
+  const guardrails = await ExperimentModel.find(
+    {
+      organization,
+      guardrails: metricId,
+    },
+    cols
+  );
+  guardrails.forEach((exp) => {
+    experiments.push({
+      id: exp.id,
+      name: exp.name,
+    });
+  });
+
+  // Using as an activation metric
+  const activations = await ExperimentModel.find(
+    {
+      organization,
+      activationMetric: metricId,
+    },
+    cols
+  );
+  activations.forEach((exp) => {
+    experiments.push({
+      id: exp.id,
+      name: exp.name,
+    });
+  });
+
+  return uniqBy(experiments, "id");
+}
+
+export async function getExperimentByIdea(
+  organization: string,
+  idea: IdeaDocument
+): Promise<ExperimentInterface | null> {
+  const experiment = await ExperimentModel.findOne({
+    organization,
+    ideaSource: idea.id,
+  });
+
+  return experiment ? toInterface(experiment) : null;
+}
+
+export async function getExperimentsToUpdate(
+  ids: string[]
+): Promise<Pick<ExperimentInterface, "id" | "organization">[]> {
+  const experiments = await ExperimentModel.find(
+    {
+      datasource: {
+        $exists: true,
+        $ne: "",
+      },
+      status: "running",
+      autoSnapshots: true,
+      nextSnapshotAttempt: {
+        $exists: true,
+        $lte: new Date(),
+      },
+      id: {
+        $nin: ids,
+      },
+    },
+    {
+      id: true,
+      organization: true,
+    }
+  )
+    .limit(100)
+    .sort({ nextSnapShotAttempt: 1 });
+  return experiments.map(toInterface);
```

Comment:
The toInterface method is meant to deal with entire experiment objects, but this query and a few below it only select a subset of fields.

Reply:
Updated to simply `return experiments`